pkg/controller/garbagecollector/memcachediscovery/client.go

```diff
@@ -0,0 +1,185 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package memcachediscovery includes a Client which is a CachedDiscoveryInterface.
+package memcachediscovery
+
+import (
+	"sync"
+
+	"github.com/emicklei/go-restful-swagger12"
+	"github.com/go-openapi/spec"
+	"github.com/golang/glog"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	restclient "k8s.io/client-go/rest"
+)
+
+// Client can Refresh() to stay up-to-date with discovery information. Before
+// this is moved some place where it's easier to call, it needs to switch to a
+// watch interface. Right now it will poll anytime Refresh() is called.
+type Client struct {
+	delegate discovery.DiscoveryInterface
+
+	lock                   sync.RWMutex
+	groupToServerResources map[string]*metav1.APIResourceList
+	groupList              *metav1.APIGroupList
+}
+
+var _ discovery.CachedDiscoveryInterface = &Client{}
+
+// ServerResourcesForGroupVersion returns the supported resources for a group and version.
+func (d *Client) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	if rl, ok := func() (*metav1.APIResourceList, bool) {
+		d.lock.RLock()
+		defer d.lock.RUnlock()
+		cachedVal, ok := d.groupToServerResources[groupVersion]
+		return cachedVal, ok
+	}(); ok {
+		return rl, nil
+	}
+
+	liveResources, err := d.delegate.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		glog.V(3).Infof("skipped caching discovery info due to %v", err)
+		return nil, err
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.groupToServerResources[groupVersion] = liveResources
+	return liveResources, nil
+}
+
+// ServerResources returns the supported resources for all groups and versions.
+func (d *Client) ServerResources() ([]*metav1.APIResourceList, error) {
+	apiGroups, err := d.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+	groupVersions := metav1.ExtractGroupVersions(apiGroups)
+	result := []*metav1.APIResourceList{}
+	for _, groupVersion := range groupVersions {
+		resources, err := d.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, resources)
+	}
+	return result, nil
+}
+
+func (d *Client) ServerGroups() (*metav1.APIGroupList, error) {
+	if g, ok := func() (*metav1.APIGroupList, bool) {
+		d.lock.RLock()
+		defer d.lock.RUnlock()
+		return d.groupList, d.groupList != nil
+	}(); ok {
+		return g, nil
+	}
+
+	liveGroups, err := d.delegate.ServerGroups()
```

Comment:
What causes empty lists?

This is going to be updated every 30s so it should be ok if it forgets stuff for 30s